### PR TITLE
Initial Support ticket entity type implementation

### PR DESCRIPTION
**Pull request by jeremyandrews**
A work in progress for creating a bundleable "Support ticket" entity (bundles are "Support ticket types"). This is essentially a stripped down Node, which will then gain Support ticket specific information once the basics are fully working.

### CODE REVIEW

**Comment by jeremyandrews** [comment]
Known problem:
After creating a Support Ticket Type, when trying to create a Support Ticket of that type I get the following error:

```
The website encountered an unexpected error. Please try again later.
Recoverable fatal error: Argument 1 passed to Drupal\Core\Access\AccessResult::orIf() must implement interface Drupal\Core\Access\AccessResultInterface, boolean given, called in /var/www/html/core/lib/Drupal/Core/Entity/EntityAccessControlHandler.php on line 83 and defined in Drupal\Core\Access\AccessResult->orIf() (line 370 of core/lib/Drupal/Core/Access/AccessResult.php).
Drupal\Core\Access\AccessResult->orIf(1)
Drupal\Core\Entity\EntityAccessControlHandler->access(Object, 'view', 'en', Object, 1)
Drupal\support_ticket\SupportTicketAccessControlHandler->access(Object, 'view', 'en', Object, 1)
Drupal\support_ticket\Entity\SupportTicket->access('view', Object, 1)
Drupal\Core\Entity\EntityAccessCheck->access(Object, Object, Object)
call_user_func_array(Array, Array)
Drupal\Core\Access\AccessManager->performCheck('access_check.entity', Object)
Drupal\Core\Access\AccessManager->check(Object, Object, Object, 1)
Drupal\Core\Access\AccessManager->checkRequest(Object, Object, 1)
Drupal\Core\Routing\AccessAwareRouter->checkAccess(Object)
Drupal\Core\Routing\AccessAwareRouter->matchRequest(Object)
Symfony\Component\HttpKernel\EventListener\RouterListener->onKernelRequest(Object, 'kernel.request', Object)
Drupal\Component\EventDispatcher\ContainerAwareEventDispatcher->dispatch('kernel.request', Object)
Symfony\Component\HttpKernel\HttpKernel->handleRaw(Object, 1)
Symfony\Component\HttpKernel\HttpKernel->handle(Object, 1, 1)
Drupal\Core\StackMiddleware\Session->handle(Object, 1, 1)
Drupal\Core\StackMiddleware\KernelPreHandle->handle(Object, 1, 1)
Drupal\page_cache\StackMiddleware\PageCache->pass(Object, 1, 1)
Drupal\page_cache\StackMiddleware\PageCache->handle(Object, 1, 1)
Drupal\Core\StackMiddleware\ReverseProxyMiddleware->handle(Object, 1, 1)
Drupal\Core\StackMiddleware\NegotiationMiddleware->handle(Object, 1, 1)
Stack\StackedHttpKernel->handle(Object, 1, 1)
Drupal\Core\DrupalKernel->handle(Object)
```

I assume this is from stripping out Node Access grants incorrectly -- my aim was to let OG enforce ticket access, instead.

**Comment by jeremyandrews** [comment]
I tracked down the error, fixed here (and added a question in the code for @LionsAd or @catch56 to be sure I'm returning the correct default as compared to the core node module): 
https://github.com/jeremyandrews/support/commit/e12597d1e63511091af25bbb7f226d69aa0daa33

**Comment by jeremyandrews** [comment]
This is functional enough, I'd like to see this reviewed and hopefully merged -- as we've got multiple people involved it'd be helpful to have this in the master repo at this point.

**Comment by jeremyandrews** [comment]
Merged fixes/cleanup from Brandon.

File: modules/support_ticket/src/PageCache/DenySupportTicketPreview.php [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+use Drupal\Core\PageCache\ResponsePolicyInterface;
+use Drupal\Core\Routing\RouteMatchInterface;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpFoundation\Response;
+
+/**
+ * Cache policy for support ticket preview page.
+ *
+ * This policy rule denies caching of responses generated by the
+ * entity.support_ticket.preview route.
```
Comment (fabianx-ai) [inline modules/support_ticket/src/PageCache/DenySupportTicketPreview.php]: This does not hurt, but can also be easier done by using:

no_cache: TRUE

on the route.

@catch56 Thoughts?

File: modules/support_ticket/templates/support-ticket.html.twig [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+ * @see template_preprocess_support_ticket()
+ *
+ * @todo Remove the id attribute (or make it a class), because if that gets
+ *   rendered twice on a page this is invalid CSS for example: two lists
+ *   in different view modes.
+ *
+ * @ingroup themeable
+ */
+#}
+<article{{ attributes }}>
```
Comment (fabianx-ai) [inline modules/support_ticket/templates/support-ticket.html.twig]: Is there anything support ticket specific in this template?

Else we can make use of twigs inheritance feature, which would save some code duplication.

File: modules/support_ticket/src/PageCache/DenySupportTicketPreview.php (hunk 1 above)
Comment (jeremyandrews) [inline modules/support_ticket/src/PageCache/DenySupportTicketPreview.php]: That makes sense to me -- should I also contribute a patch for the Node module which is where I copied this code from?

File: modules/support_ticket/templates/support-ticket.html.twig (hunk 2 above)
Comment (jeremyandrews) [inline modules/support_ticket/templates/support-ticket.html.twig]: This is what the node module does -- I didn't really understand it, but I copied it. Do I just delete it, or do I need to do something else to make use of twigs inheritance feature?

File: modules/support_ticket/src/PageCache/DenySupportTicketPreview.php (hunk 1 above)
Comment (catch56) [inline modules/support_ticket/src/PageCache/DenySupportTicketPreview.php]: Yes I think no_cache: TRUE would be better in both cases.

File: modules/support_ticket/templates/support-ticket.html.twig (hunk 2 above)
Comment (catch56) [inline modules/support_ticket/templates/support-ticket.html.twig]: Is support_ticket handling display_submitted and other things from node as well?

File: modules/support_ticket/config/optional/field.field.support_ticket.ticket.field_assigned_to.yml [hunk 3]
```diff
@@ -0,0 +1,30 @@
+uuid: 9d53bedd-ebbc-4d6a-ba32-50d2ddd45c83
```
Comment (catch56) [inline modules/support_ticket/config/optional/field.field.support_ticket.ticket.field_assigned_to.yml]: iirc the uuid needs to be stripped from these.

Comment (jeremyandrews) [inline modules/support_ticket/config/optional/field.field.support_ticket.ticket.field_assigned_to.yml]: I wondered about that -- these are from when I export configuration. Should I completely strip the uuid line in all config files, or does some sort of filler go there? (I think I saw something like "uuid: UUID" or something like that in some modules).

File: modules/support_ticket/templates/support-ticket.html.twig (hunk 2 above)
Comment (jeremyandrews) [inline modules/support_ticket/templates/support-ticket.html.twig]: Yes, all the twig template files are copied from node -- I basically just replaced node with support_ticket in these.

File: modules/support_ticket/src/Entity/SupportTicket.php [hunk 4]
```diff
@@ -0,0 +1,468 @@
...
+  /**
+   * {@inheritdoc}
+   */
+  public static function preDelete(EntityStorageInterface $storage, array $entities) {
+    parent::preDelete($storage, $entities);
+
+    // Ensure that all support_tickets deleted are removed from the search index.
+    if (\Drupal::moduleHandler()->moduleExists('search')) {
+      foreach ($entities as $entity) {
+        search_index_clear('support_ticket_search', $entity->stid->value);
```
Comment (catch56) [inline modules/support_ticket/src/Entity/SupportTicket.php:130]: Not an issue for here, but I wonder a bit whether search indexing could re-use cache metadata.

File: modules/support_ticket/src/Form/SupportTicketPreviewForm.php [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+use Drupal\Core\Entity\EntityManagerInterface;
+use Drupal\Core\Form\FormBase;
+use Drupal\Core\Form\FormStateInterface;
+use Drupal\Core\Url;
+use Symfony\Component\DependencyInjection\ContainerInterface;
+
+/**
+ * Contains a form for switching the view mode of a support ticket during preview.
+ */
+class SupportTicketPreviewForm extends FormBase implements ContainerInjectionInterface {
```
Comment (catch56) [inline modules/support_ticket/src/Form/SupportTicketPreviewForm.php]: Is this worth implementing?

File: modules/support_ticket/src/PageCache/DenySupportTicketPreview.php (hunk 1 above)
Comment (catch56) [inline modules/support_ticket/src/PageCache/DenySupportTicketPreview.php]: Or max_age = 0 on the controller.

Comment (jeremyandrews) [inline modules/support_ticket/src/PageCache/DenySupportTicketPreview.php]: no_cache: TRUE seems the most explicit to me -- updated the code here, will submit a patch for core (and check to see if this pattern occurs in other modules too).

File: modules/support_ticket/config/optional/field.field.support_ticket.ticket.field_assigned_to.yml (hunk 3 above)
Comment (jeremyandrews) [inline modules/support_ticket/config/optional/field.field.support_ticket.ticket.field_assigned_to.yml]: It's documented here: https://www.drupal.org/node/2087879  -- I'm just removing the uuid's.

File: modules/support_ticket/src/Form/SupportTicketPreviewForm.php (hunk 5 above)
Comment (jeremyandrews) [inline modules/support_ticket/src/Form/SupportTicketPreviewForm.php]: It seems to not be used at all -- removing it.

File: modules/support_ticket/templates/support-ticket.html.twig (hunk 2 above)
Comment (jeremyandrews) [inline modules/support_ticket/templates/support-ticket.html.twig]: I tried to use inheritance, but ultimately had to add more code not less so there are now support_ticket classes in the viewed ticket body.

File: modules/support_ticket/config/install/system.action.support_ticket_delete_action.yml [hunk 6]
```diff
@@ -0,0 +1,9 @@
+id: support_ticket_delete_action
+label: 'Delete support ticket'
+status: true
+langcode: en
+type: support_ticket
+plugin: support_ticket_delete_action
```
Comment (fabianx-ai) [inline modules/support_ticket/config/install/system.action.support_ticket_delete_action.yml:6]: @generic entity_delete_action

File: modules/support_ticket/config/install/system.action.support_ticket_lock_action.yml [hunk 7]
```diff
@@ -0,0 +1,9 @@
+id: support_ticket_lock_action
+label: 'Lock support ticket'
+status: true
+langcode: en
+type: support_ticket
+plugin: support_ticket_lock_action
```
Comment (fabianx-ai) [inline modules/support_ticket/config/install/system.action.support_ticket_lock_action.yml:6]: @generic

File: modules/support_ticket/config/install/system.action.support_ticket_delete_action.yml [hunk 8]
```diff
@@ -0,0 +1,9 @@
+id: support_ticket_delete_action
+label: 'Delete support ticket'
```
Comment (fabianx-ai) [inline modules/support_ticket/config/install/system.action.support_ticket_delete_action.yml:2]: As step 2, should add configurable labels, or rather:

Start with 'Delete entity' here, then override with 'Delete %bundle_name' later.

File: modules/support_ticket/config/install/system.action.support_ticket_lock_action.yml (hunk 7 above)
Comment (jeremyandrews) [inline modules/support_ticket/config/install/system.action.support_ticket_lock_action.yml:6]: This is actually a support-ticket-specific feature. I got rid of "promote" and "sticky", but added "lock" (which is a way to lock a ticket so only admins can modify them). As this is not a concept in a node, I don't think it's actually generic.

File: modules/support_ticket/config/schema/support_ticket.schema.yml [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+  label: 'Support ticket settings'
+  mapping:
+    use_admin_theme:
+      type: boolean
+      label: 'Use admin theme when editing or creating tickets'
+
+support_ticket.type.*:
+  type: config_entity
+  label: 'Support ticket type'
+  mapping:
```
Comment (fabianx-ai) [inline modules/support_ticket/config/schema/support_ticket.schema.yml:14]: @generic I think we can inherit the schema of the generic entity here.

File: modules/support_ticket/src/Access/SupportTicketAddAccessCheck.php [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+   *   there exists at least one support_ticket type for which the user may create
+   *   a support_ticket.
+   *
+   * @return string
+   *   A \Drupal\Core\Access\AccessInterface constant value.
+   */
+  public function access(AccountInterface $account, SupportTicketTypeInterface $support_ticket_type = NULL) {
+    $access_control_handler = $this->entityManager->getAccessControlHandler('support_ticket');
+    // If checking whether a support_ticket of a particular type may be created.
+    if ($account->hasPermission('administer support tickets types')) {
```
Comment (fabianx-ai) [inline modules/support_ticket/src/Access/SupportTicketAddAccessCheck.php:56]: @generic This super permission is the only thing that is hardcoded in here:

Either we use:

t('administer %entity types')

or

we push that functionality generically into the access control handler, where it IMHO belongs.

File: modules/support_ticket/src/Access/SupportTicketAddAccessCheck.php [hunk 11]
```diff
@@ -0,0 +1,73 @@
...
+   *   The currently logged in account.
+   * @param \Drupal\support_ticket\SupportTicketTypeInterface $support_ticket_type
+   *   (optional) The support_ticket type. If not specified, access is allowed if
+   *   there exists at least one support_ticket type for which the user may create
+   *   a support_ticket.
+   *
+   * @return string
+   *   A \Drupal\Core\Access\AccessInterface constant value.
+   */
+  public function access(AccountInterface $account, SupportTicketTypeInterface $support_ticket_type = NULL) {
```
Comment (fabianx-ai) [inline modules/support_ticket/src/Access/SupportTicketAddAccessCheck.php:53]: @generic GenericEntityTypeInterface

File: modules/support_ticket/src/Access/SupportTicketAddAccessCheck.php [hunk 12]
```diff
@@ -0,0 +1,73 @@
...
+    $access_control_handler = $this->entityManager->getAccessControlHandler('support_ticket');
```
Comment (fabianx-ai) [inline modules/support_ticket/src/Access/SupportTicketAddAccessCheck.php:54]: @generic Need to pass the 'support_ticket' information into the class.

Likely on __construct

File: modules/support_ticket/src/Access/SupportTicketPreviewAccessCheck.php [hunk 13]
```diff
@@ -0,0 +1,60 @@
...
+   *
+   * @param \Drupal\Core\Session\AccountInterface $account
+   *   The currently logged in account.
+   * @param \Drupal\support_ticket\SupportTicketInterface $support_ticket_preview
+   *   The support_ticket that is being previewed.
+   *
+   * @return string
+   *   A \Drupal\Core\Access\AccessInterface constant value.
+   */
+  public function access(AccountInterface $account, SupportTicketInterface $support_ticket_preview) {
```
Comment (fabianx-ai) [inline modules/support_ticket/src/Access/SupportTicketPreviewAccessCheck.php:50]: @generic Except for the Interface and the hardcoded 'support_ticket', this is generic functionality.

File: modules/support_ticket/config/schema/support_ticket.schema.yml [hunk 14]
```diff
@@ -0,0 +1,86 @@
...
+support_ticket.settings:
+  type: config_object
+  label: 'Support ticket settings'
+  mapping:
+    use_admin_theme:
+      type: boolean
+      label: 'Use admin theme when editing or creating tickets'
+
+support_ticket.type.*:
+  type: config_entity
```
Comment (fabianx-ai) [inline modules/support_ticket/config/schema/support_ticket.schema.yml:12]: @generic Yes, we can:

http://hojtsy.hu/files/ConfigSchemaCheatSheet1.5.pdf

Would use:

type: generic_entity

here or something like that.

Comment (fabianx-ai) [inline modules/support_ticket/config/schema/support_ticket.schema.yml:12]: We could already inherit from node now, but given we want to make this generic, not sure that makes sense.

Also as a generic thing:

We can do the generalization in little steps.

**Pull request merged**